### PR TITLE
Feat/dkg contract interaction

### PR DESCRIPTION
**Pull request by soju-drinker**
### Description

This PR adds the ability for the Signer to interact with Clarity contracts

### Applicable issues
- needed for https://github.com/stacks-network/stacks-blockchain/issues/3971

### Checklist
- [x] Read only contract calls
- [x] Submit tx contract calls
- [x] Simulate contract call responses in tests

### CODE REVIEW

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:75]: Just nitpicky but I prefet this being called StacksContractable :P

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:247]: we will have to turn this probably into some sort of retry with backoff logic and never panic.

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:74]: Really nitpicky. Can ignore if you like, but think maybe something like "StacksContractCallable" is better. I can be easily convinced otherwise though :P

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:282]: Typo. Should be Content-Type

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:288]: Mind running clippy on this PR? 
`cargo clippy -p stacks-signer --tests --no-deps --all-features`

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs:258]: Would you mind updating this to return a result. And then move the "FIXME" to the panic to the external calling of this function? Think its better for the traits to never panic and leave it to the caller to panic.

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:237]: I would maybe make client part of StacksClient struct and pass it in here. This way we can just keep reusing the same one/not create a new connection every time.

File: stacks-signer/Cargo.toml
Comment (jferrant) [inline stacks-signer/Cargo.toml:45]: I do believe Jude has concerns about using reqwest because of unsafe code in the blocking call. We may need to hunt around for a different crate :(

Comment (jferrant) [reply stacks-signer/Cargo.toml:45]: I would also be happy with allowing this through for now and making a ticket to repalce it with some alternative. I think the HTTP code throughout stacks is being pulled into a lib and maybe we could reuse that?

Comment (jcnelson) [reply stacks-signer/Cargo.toml:45]: > I do believe Jude has concerns about using reqwest because of unsafe code in the blocking call. We may need to hunt around for a different crate :(

I have many, many concerns about `reqwest` -- in particular, the god-awful dependencies it pulls in.

If you're only doing blocking calls, I'd recommend `ureq`.  No need to import the whole and run the whole of `tokio` just to read and write data on a socket.

> I really really really do not want to have to implement a custom http client.

Oh but it's so much fun! ;)

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs:74]: Can I ask what the trait definition here is for?  Are there two or more implementations?  If not, then perhaps the trait can be dropped altogether.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:185]: Can this just return an `Err(..)` instead of panicking?

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:156]: Let's not `panic` if we don't need to :)  This should be a `Result`.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:191]: There's `StacksMessageCodec::serialize_to_vec()` which converts `T: StacksMessageCodec` into `Vec<u8>`:

```rust
// N.B. this error variant needs to exist
let tx = tx_signer.get_x().ok_or(Error::TransactionSignatureFailure)?;
let buf = tx.serialize_to_vec();
```

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs:74]: A lot of the trait's functions take the same arguments over and over (e.g. `http_origin`, `chain_id`, `tx_version`).  I think these are values that the trait implementation can just supply.  Why would a caller ever need to supply different values for the same trait implementation instantiation?

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:146]: It might make more sense to have one function for non-sponsored transactions, and one for sponsored transactions.  Then, the caller won't have to supply `None` gratuitously.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:236]: I don't know if this is ever going to be a concern, but if you're registering a BNS name, then you'd also need to supply an `Attachment` which contains the name's zonefile data.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs:258]: Seconded.  Let's not panic if we can help it.  Panicking is the library consumer's job :)

File: Cargo.lock
Comment (jcnelson) [inline Cargo.lock:1740]: When you switch to `ureq`, please be sure to only use the minimum features required to solve the problem (e.g. JSON support and that's it).  No need for cookies or TLS -- the Stacks node supports neither anyway.

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs]: I am not entirely sure what this means XD Forgive me...

File: stacks-signer/Cargo.toml
Comment (jferrant) [reply stacks-signer/Cargo.toml:45]: @soju-drinker pulled out the ureq suggestions into a separate issue - https://github.com/stacks-network/stacks-blockchain/issues/3995

Hopefully we can get this merged asap to build off of (prevent massive PR)

Comment (jcnelson) [reply stacks-signer/Cargo.toml:45]: Just to be clear, this is going to need to be addressed before a merge to `master`.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:64]: Can serialization fail?  Or is this just for signature failure?

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:115]: We have specific containers for contract names, function names, and function arguments, since they have to use limited character sets and since not all strings are valid.  You should use `ContractName` and `ClarityName` here instead of a bare `&str`.  Also, `function_args` should be `&[clarity::vm::Value]` in both cases.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:256]: These can panic.  Yes, it's bad rust practice, but it is what it is today.  `::from(x)` just does `::try_from(x).unwrap()` for `ClarityName` and `ContractName`.  This is why this trait interface should use `ContractName` and `ClarityName`.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:284]: This should not panic.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:286]: Please remove all panicking unwraps and expects.  This method is already fallible, so it should just return an `Err(..)` if the network request fails for some reason.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:308]: This can fail if `function_args` does not contain serialized `clarity::vm::Value`s.  This interface should instead take `&[clarity::vm::Value]` for `function_args`.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:311]: Same problem here -- `contract_name` should be a `ContractName` and `function_name` should be a `ClarityName` in the interface, because this RPC will fail if you pass an invalid contract name or function name string.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [inline stacks-signer/src/stacks_client.rs:439]: You might get a 400 or a 404 as well.  Can you test these conditions too?

**Review by jcnelson — CHANGES_REQUESTED** [review]
This PR is on the right track, but it needs a few changes:

* It needs to use `ContractName`, `ClarityName`, and `clarity::vm::Value` instead of bare `&str`s
* The `StacksContractCallable` trait may not need to exist

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs:420]: Done.

**Comment by jferrant** [comment]
> This PR is on the right track, but it needs a few changes:
> 
>     * It needs to use `ContractName`, `ClarityName`, and `clarity::vm::Value` instead of bare `&str`s
> 
>     * The `StacksContractCallable` trait may not need to exist

Done :)

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [inline stacks-signer/src/stacks_client.rs:109]: Gotta fix.

Comment (jferrant) [reply stacks-signer/src/stacks_client.rs]: Fixed.

**Review by jferrant — APPROVED** [review]
LGTM

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs]: Quick note, can you create a valid "StacksAddress" from the pox contract address? My initial attempts have failed. The reason I changed this to String was to accommodate this. However, perhaps I am just doing it wrong?

Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs]: Sure -- `QualifiedContractIdentifier::parse("ST000000000000000000002AMW42H.pox-3").unwrap()` should work.

File: stacks-signer/src/stacks_client.rs
Comment (jcnelson) [reply stacks-signer/src/stacks_client.rs]: Ah, so there's this storage system in the Stacks node called Atlas  It's a hold-over from the Blockstack / Stacks 1.0 days.  Basically, a smart contract can emit specially-crafted Atlas events which contain the hash of a chunk of data.  The Atlas system tracks these hash announcements, and then communicates with other Stacks nodes to see if they have a chunk of data which hashes to it.  Eventually, all nodes get a copy of this "attached" data.  We use it in BNS to tie a BNS name to a zone file, as well as other off-chain data like Nostr public keys and such.

I don't think the signer is ever going to use Atlas, so it's not a big deal if you're not going to support attachments.  However, if this module becomes more generic and gets spun off into its own crate, it might become important later to add support.

File: stacks-signer/src/stacks_client.rs
Comment (jferrant) [reply stacks-signer/src/stacks_client.rs]: Yes sorry I wasn't clear/ I am saying that does not work 
 'called `Result::unwrap()` on an `Err` value: InvalidContractAddress("\"ST000000000000000000002AMW42H")',

Comment (jferrant) [reply stacks-signer/src/stacks_client.rs]: Sorry, I was not clear. This is what I already tried and it results in 
```thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Runtime(ParseError("Invalid principal literal: base58ck checksum 0x51104e95 does not match expected 0x31287a31"), None)', stacks-signer/src/stacks_client.rs:315:83```

This is the same result as trying to parse ST000000000000000000002AMW42H as a StacksAddress.

**Review by xoloki — APPROVED** [review]
LGTM.  More tests would be nice, especially if they weren't mocked, but this is fine for now.

**Review by jcnelson — APPROVED** [review]

**Pull request merged**
